File: src/algorithms/search-algorithms/binary-search/binary-search.js
const binarySearchRecursive = (array, left, right, elem) => {
    if (right >= left) {
        let middle = left + Math.floor((right - left) / 2) ;

        if (array[middle] === elem) {
            return middle;
        }

        if (array[middle] > elem) {
            return binarySearch(array, left, mid - 1, elem);
        }

        return binarySearch(array, mid + 1, right, elem);
    }

    return -1;
}


const array = [1, ,2 ,3 ,4 ,5, 6, 7];
const elem = 6;
const index = binarySearchRecursive(array, 0, array.length, elem);
console.log(index);

File: src/algorithms/search-algorithms/binary-search/binary-search.go
package binarySearch

func BinarySearchRecursive(array []int, left int, right int, elem int) int {
	if right >= left {
		middle := left + (right-left)/2
		if array[middle] == elem {
			return middle
		}

		if array[middle] > elem {
			return BinarySearchRecursive(array, left, middle-1, elem)
		}

		return BinarySearchRecursive(array, middle+1, right, elem)
	}
	return -1
}

func BinarySearchIterative(array []int, elem int) int {
	left := 0
	right := len(array) - 1
	var middle int

	for right >= left {
		middle = left + (right-left)/2

		if array[middle] == elem {
			return middle
		}

		if array[middle] > elem {
			right = middle - 1
		} else {
			left = middle + 1
		}
	}

	return -1
}
